### ISSUE TITLE
Reset Form No Longer Works With Custom Filtering (v3)

### ISSUE THREAD

**Issue by jeffconklin** [fooplugins/FooTable#522]
Hello,

I have been scratching my head for a bit with my local development trying to figure out why my custom filter (similar to http://fooplugins.github.io/FooTable/docs/examples/advanced/filter-dropdown.html) does not allow me to "Cancel" the filtering after switching a dropdown. For example..
- Go to the Filter Dropdown example: http://fooplugins.github.io/FooTable/docs/examples/advanced/filter-dropdown.html
- Change the dropdown
- Note that the Magifying Glass did not switch to the X, and clicking on it will not reset the filter.
- Next, add something into the textbox.
- Note that the cancel icon is displayed, and clicking this "cancels" the current filtering, resets the textbox, and resets the select.

I am using this style custom filtering for a bit before 3.1.0 and I never had this issue.

**Comment by steveush** [comment]
Hey Jeff,

I've just checked in a fix for this into the develop branch. I checked in the [develop compiled folder](https://github.com/fooplugins/FooTable/tree/develop/compiled) with the updated changes so you can grab the JS directly from there instead of having to build it. Please test and let me know if it works as expected, I tested it my side and everything seems right now.

Thanks

**Comment by jeffconklin** [comment]
Hi @steveush,

I grabbed from the compiled directory and now the "X" is always displayed. 
![2016-09-15-ft-admin](https://cloud.githubusercontent.com/assets/6152325/18563811/137e2dae-7b57-11e6-81d2-9247de2ddfa7.png)

The only thing I would like to add is that I have a default filter loaded when the table loads...
`table data-filter-filters='[{"name":"tblfilter_a","query":"Active","columns":["tblfilter_a"]}]'`

Clicking the X will remove any applied filtering, but will never go back to the mag icon since I have that default enabled anytime that you are not actual filtering something different..

For a quick test, I removed the default and it displayed properly. 
![2016-09-15-ft-admin-withoutdefault](https://cloud.githubusercontent.com/assets/6152325/18564011/bb2c9f72-7b57-11e6-80a8-bd10d2f36758.png)

I am going to revert back to your release. I prefer to be in sync with your release over having some test code mixed in.

Thank you for quick reply as well.

Edit: Wanted to add that both screen shots are on page load, after initialization, before I interact with/click anything dealing with the plugin.

**Comment by steveush** [comment]
OK, will check it out but this is the expected behavior in my eyes, there is a filter applied (through the data attributes) so the icon is the X. What I think is needed here on top of the current change is a way to specify if a filter should be a permanent filter that does not get cleared or counted when interacting with the X button.

**Comment by steveush** [comment]
Just wanted to add that I think you are seeing this behavior now as in 3.1.0 there is the following fix:
- Fixed an issue with filtering not applying correctly when filters were supplied through the options.

So previously the filter supplied through the data attributes was not being applied correctly, the 3.1.0 release now fixes that issue so now you are seeing the clear button on init as the filter is actually been applied now.

EDIT: The behavior of the develop branch is now identical to the 3.0.x releases except that now the filters supplied through the options are being properly applied.

**Comment by jeffconklin** [comment]
Hmm... Well that definitely makes sense... a filter is active, the X/Cancel should also be active. Within my pages I had to create a hack to get things to work how I want because this seemed "off". But now changing this to allow it to work "properly" is a pretty big deal for the user experience. Especially based on how this has been working for some time now.

Please review this example: http://output.jsbin.com/wakuvi/5

Once you address the current X/Cancel issue I originally reported, the X/Cancel will be displayed onload since a "default" filter has been applied. But to the user there is no filter... it is just "the page". Certainly confusing. Especially since other pages that do not have default filters will include the search icon to start, and the search icon after the X/Cancel is clicked. That will certainly be reported as a bug by my users.

I guess the next step is to understand what my options are here...

A) How can I set an actual default filter that is ALWAYS displayed to the user onload. It is also applied after the  X/Cancel. In both situations, on Cancel, the Search icon should be displayed. That would allow my situation to "just work", and be more predictable/controllable.
B) How can I add/remove rows/data without having to destroy/recreate the table with each filter change? I do not think you have yet created a way to add/remove rows... maybe I am wrong?

This certainly changes a lot for our pages. Any direction would be great.

Thanks!

**Comment by jeffconklin** [comment]
> What I think is needed here on top of the current change is a way to specify if a filter should be a permanent filter that does not get cleared or counted when interacting with the X button.

Somehow missed that previous post. But yes, that seems to be a very logical solution for an "onload" filter. Being able to choose either would certainly add some power here.

**Comment by jeffconklin** [comment]
Talked this over a bit more with colleagues ... being able to declare a "default" filter would be a great start and the best solution for many of our tables. But also being able to just grab new data from the server in an easier way would be sick. Right now I can destroy the table, request more data, and reinitialize. But that means everything visually goes away for the user... even for a second. Not a good experience for the user.

Thanks again for everything.

**Comment by steveush** [comment]
Hey Jeff, check out #512 for a nice simple way of loading new rows into a pre-existing table using the `FooTable.Rows#load(array, bool)` method. If you take a look at how this method works under the hood you can also see how you can manipulate the rows yourself if needed. I'm making the changes at the moment on the develop branch to allow for a "hidden" option for filters. When set to true they are always applied to a table, can not be cleared unless removed using the `FooTable.Filtering#removeFilter()` method and they will not effect the default UI buttons.

**Comment by steveush** [comment]
The above mentioned `hidden` option for a filter is now available in 3.1.1 which has just been pushed. To use your original example it would look like the below to make use of the option:

``` html
<table data-filter-filters='[{"name":"tblfilter_a","query":"Active","columns":["tblfilter_a"],"hidden":true}]' />
```

EDIT: Forgot to mention I threw together a small example of loading rows that can be found [here](http://fooplugins.github.io/FooTable/docs/examples/advanced/loading-rows.html) in the docs.

Thanks

**Comment by jeffconklin** [comment]
Thanks... I will certainly check out 3.1.1 soon. And your loading example page looks like it will be a great guide... I will jump on that soon.

Your explanation of the "hidden":true feature seems pretty powerful. And this is available and acts the same in both data-filter-filters and through the addFilter call? If so, that will be an easy change on my end. Again my goal is to have a default filter, remove that filter when someone wants to actually run a real filter, and add the default filter back in again on clear. All with the Clear button functioning for the user in a way that makes logical sense to them. Seems like "hidden":true might do the trick.

Thanks again, @steveush. Your hard work is certainly appreciated by many people.

**Comment by steveush** [comment]
Hi Jeff,

Probably the easiest way to think of the `hidden` option is that once a filter is set to hidden it will always be applied, even after calling the `FooTable.Filtering#clear()` method or applying additional filters. This means all your custom filtering code should not have to change at all, any hidden filters will just keep being applied no matter what you do until you manually call `FooTable.Filtering#removeFilter(name)`. Also to answer your question about if you can supply this option through the [`FooTable.Filtering#addFilter()`](http://fooplugins.github.io/FooTable/docs/jsdocs/FooTable.Filtering.html#addFilter) method, yes you can. The arguments have been updated to include all the available `FooTable.Filter` constructor parameters.

Thanks

**Comment by jeffconklin** [comment]
Ran a quick test and can confirm that the initial load filtering is working properly, and the Search/Clear button is showing the Search icon. Any filtering beyond this point displays the X. This is working properly now.

I am having an issue when clearing the custom filter. I call removeFilter and use addFilter() to add the original/default filter back on. Though I am passing in true as the last param, the the Search/Clear button is still display the clear (X). Based on what you mentioned above, passing in "hidden":true within addFilter will work the same as being passed as a data attribute.

Also, on page load, just clicking the Search/Clear seems to not care about the default search, and "No results" is displayed in the table. I added a new issue about this ( #524 ) since this current thread is more about default filtering options.

I will try to create a jsbin example today for further, more appropriate testing.

**Comment by jeffconklin** [comment]
Hey @steveush ,

Creating a new demo outside my current dev environment allowed me to see the new, proper way to create custom filter/dropdowns. Once I started fresh using your updated filter dropdown example (http://fooplugins.github.io/FooTable/docs/examples/advanced/filter-dropdown.html), things fell into place. The main difference between your previous version this one is that the new one does not include clear: function(){}. Once I added this back in, I was able to set the default filter on clear.

Here is the working demo:  http://output.jsbin.com/vugezes

As usual, well done here. Your recent updates have been very useful. Thank you.
